presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -494,13 +494,12 @@ public boolean transitionToFailed(Throwable throwable)
         boolean failed = queryState.setIf(FAILED, currentState -> !currentState.isDone());
         if (failed) {
             log.debug(throwable, "Query %s failed", queryId);
+            session.getTransactionId().ifPresent(autoCommit ? transactionManager::asyncAbort : transactionManager::fail);
```

Comment:
if you can race here, the above failureCause compare and set also needs to be here too